### DISCUSSION TITLE
Managing traffic question

### DISCUSSION THREAD

**Discussion by userH1213** [InterDigitalInc/AdvantEDGE#360] [Q&A]
Hi,
 
  I'm working on a project and I have some questions. 
1- What scheduling scheme is used in AdvantEDG when it comes to sending packets to the servers(routing packets) is it FIFO or something else? 
2- If I want to implement my own scheduling algorithm where should I implement it(I've checked the files in demo1,demo2 and demo3 but can't figure out where exactly) Also, where it would be represented ( poa, operators..)?

Thank you

**Comment by roymx** [comment]
Hello @userH1213,

Thanks for reaching out.

First, I would suggest reading [this section](https://interdigitalinc.github.io/AdvantEDGE/docs/overview/overview-architecture/#network-characteristics) from the documentation, it provides a very high level overview of what AdvantEDGE does in terms of controlling traffic.

Short answer to your question is: traffic scheduling is performed in the kernel.
AdvantEDGE implements a form of distributed _tc_netem_ to control the network characteristics according to the an edge scenario topology and configuration.

More specifically:
- Each pod has its own IP stack.
- AdvantEDGE's TC-engine pod consumes the active scenario and constantly re-evaluates the point-to-point network characteristics between all scenario pods according to scenario and measurements
- AdvantEDGE's TC-sidecars are injected in all scenario pods at deployment time and apply network characteristics according to TC-engine evaluations
- Each TC-sidecar uses _tc-netem_ to setup the required _ifb_ in each pod
- The above is repeated every seconds according the the network changes

The files you looked into (demo) are the scenario applications (e.g. pods), therefore you will not find anything related to the network in those as network control happens at an underneath layer. Those application just use the network (e.g. open a socket and transfer traffic) and experience network characteristics according to IP stack configuration.

It is not totally clear to me what scheduling algorithm you are referring to.

Good luck!

**Reply by userH1213** [reply]
roymx,
Thank you for your reply. To clarify more, I'm trying to understand how different scheduling algorithms affect packet flow from ue-app to cloud-app. By scheduling algorithms I mean FIFO, Bottleneck Fairness, Dominant Resource Fairness. 

Thank you

**Reply by userH1213** [reply]
Hi again,

I'm sorry if my questions were not clear enough.
- Let's say we have ue-1 and ue-2 and they both sent requests to be processed at some server in the cloud. Their requests shall be routed to the server to be processed. My question is which element is responsible for routing the packets to the server is it Operator, Poa or Edge/Fog?  Which ue's packets are going to be  sent/routed to the server first, is it ue-1's or ue-2's , I mean based on what scheduling algorithm FIFO,  CFS( the kernel's algorithm) or something else? 

-If we modify routers.go in demo1 for example are we going to be able to control the resources given to the packets and in which order they will be sent out to the server ?

-After creating a scenario with AdvantEDGE, I understand that there are containers being created.. how can I access those containers to implement more specific elements (I'm just starting with Docker and Kubernetes so maybe my question sounds not so smart:) )

Thank you for the help in advance.
